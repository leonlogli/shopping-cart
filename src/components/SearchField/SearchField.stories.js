import React from 'react'

import SearchField from './SearchField'

export default {
  title: 'SearchField',
}

export const searchField = () => <SearchField />

export const open = () => <SearchField open />

export const naked = () => <SearchField open naked />
