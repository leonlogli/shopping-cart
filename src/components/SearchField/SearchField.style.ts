import styled, { css } from 'styled-components'
import InputBase from '@material-ui/core/InputBase'
import { fade } from '@material-ui/core/styles'

const Input = styled(InputBase)`
  color: inherit;
  width: 100%;
  transform: scale(0);

  .MuiInputBase-input {
    width: 100%;

    ${({ theme }) => css`
      padding: ${theme.spacing(1, 0.5, 1, 2.5)};
      padding-right: 52px;
    `}
  }
`

const ContentBox = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;

  position: relative;
  border: none;
  width: 0;
  border-radius: 40px;
  border: none;
  background-color: ${fade('#fff', 0.15)};
  transition: none;

  ${({ theme }) => css`
    ${theme.breakpoints.up('sm')} {
      transition: width 0.3s;
    }
  `}

  border-color: ${({ theme, naked }: any) =>
    naked ? 'transparent' : theme.palette.primary.main} !important;

  &:hover {
    background-color: ${fade('#fff', 0.25)};
  }

  > button {
    margin: 0 8px;
    padding: 8px;
    right: 0;
    position: absolute;
  }
`

const SearchFieldRoot = styled.div`
  min-width: 80px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  &.showed {
    ${ContentBox} {
      width: 100%;
      border: 1px solid;

      > button,
      > button:hover {
        background: transparent;
        cursor: text;
      }
    }

    ${Input} {
      transform: scale(1);
    }
  }
`

export { SearchFieldRoot, ContentBox, Input }
export default SearchFieldRoot
