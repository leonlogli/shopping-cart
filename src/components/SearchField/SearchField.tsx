import React, { forwardRef, useRef } from 'react'
import SearchIcon from '@material-ui/icons/Search'
import { InputBaseProps } from '@material-ui/core/InputBase'

import IconButton from '@material-ui/core/IconButton'

import SearchFieldRoot, { ContentBox, Input } from './SearchField.style'

export type SearchFieldProps = InputBaseProps & {
  onSearchIconClick?: () => void
  open?: boolean
  naked?: boolean
}

/**
 * Main search field
 */
const SearchField = forwardRef<HTMLDivElement, SearchFieldProps>(
  (props, ref) => {
    const {
      className,
      id,
      onSearchIconClick,
      placeholder = 'Searchâ€¦',
      inputProps,
      open,
      naked,
      ...other
    } = props

    const contentBoxRef = useRef<HTMLElement>()

    return (
      <SearchFieldRoot
        id={id}
        ref={ref}
        className={`${className || ''} SearchField ${open ? 'showed' : ''}`}
      >
        <ContentBox ref={contentBoxRef as any} {...{ naked }}>
          <IconButton
            disableRipple
            aria-label="search"
            onClick={onSearchIconClick}
          >
            <SearchIcon color="inherit" />
          </IconButton>
          <Input
            type="search"
            placeholder={placeholder}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            spellCheck="false"
            inputProps={{
              'aria-label': placeholder,
              'aria-autocomplete': 'list',
              'aria-haspopup': 'false',
              ...inputProps,
            }}
            {...other}
          />
        </ContentBox>
      </SearchFieldRoot>
    )
  }
)

export { SearchField }
export default SearchField
