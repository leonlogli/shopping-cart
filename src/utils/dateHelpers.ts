const durationFormat = (interval: number, type: string) => {
  return `il y a ${Math.floor(interval)} ${type}${
    Math.floor(interval) > 1 ? 's' : ''
  }`
}

const fromNow = (date: Date) => {
  const seconds = Math.floor(
    ((new Date() as any) - (new Date(date) as any)) / 1000
  )

  let interval = seconds / 31536000

  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'an')
  }

  interval = seconds / 2592000
  if (Math.floor(interval) >= 1) {
    return `il y a ${Math.floor(interval)} mois`
  }

  interval = seconds / 86400
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'jour')
  }

  interval = seconds / 3600
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'heure')
  }

  interval = seconds / 60
  if (Math.floor(interval) >= 1) {
    return durationFormat(interval, 'minute')
  }

  return `Ã  l'instant`
}

/**
 * Parse the given string time to a date.Time must have `hh:mm:ss` format
 * @param time time to parse
 * @param date date to which the time will be set once parsed. If null, today's date is used
 */
const parseTimeToDate = (time: string, date?: Date) => {
  const dateToSet = date || new Date()
  const timeParts = time.split(':').map((t) => Number(t) || 0)

  dateToSet.setHours(timeParts[0], timeParts[1], timeParts[2] || 0)

  return dateToSet
}

const formatShortDate = (date: any, locale?: string) => {
  return new Date(date).toLocaleDateString(locale, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  })
}

export { fromNow, parseTimeToDate, formatShortDate }
