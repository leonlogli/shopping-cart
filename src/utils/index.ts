export * from './testHelpers'
export * from './util'
export * from './dateHelpers'
