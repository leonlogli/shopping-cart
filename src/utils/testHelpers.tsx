import React from 'react'
import { ThemeProvider } from 'styled-components'
import { mount, shallow } from 'enzyme'
import { defaultTheme } from '../styles'

const ThemeProviderWrapper = ({ children }: any) => (
  <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
)

export const shallowWithTheme = (tree: any) =>
  shallow(tree, {
    wrappingComponent: ThemeProviderWrapper,
  })

export const mountWithTheme = (tree: any) =>
  mount(tree, {
    wrappingComponent: ThemeProviderWrapper,
  })
