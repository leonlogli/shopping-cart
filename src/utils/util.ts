const toUSD = (value: number) => {
  return `${value.toLocaleString('en-US', {
    minimumFractionDigits: 2,
  })} USD`
}

export { toUSD }
