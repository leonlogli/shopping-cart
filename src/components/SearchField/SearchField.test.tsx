import React from 'react'

import SearchField from './SearchField'
import { mountWithTheme } from '../../utils'

describe('<SearchField />', () => {
  const wrapper = mountWithTheme(<SearchField />)

  it('should render children', () => {
    expect(wrapper.find('input').length).toEqual(1)
  })

  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
