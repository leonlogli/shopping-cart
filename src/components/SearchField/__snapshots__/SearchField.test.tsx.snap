// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> should match a snapshot 1`] = `
<ForwardRef>
  <styled.div
    className=" SearchField "
  >
    <div
      className="gZiGuV  SearchField "
    >
      <styled.div>
        <div
          className="iqexBz"
        >
          <WithStyles(ForwardRef(IconButton))
            aria-label="search"
            disableRipple={true}
          >
            <ForwardRef(IconButton)
              aria-label="search"
              classes={
                Object {
                  "colorInherit": "MuiIconButton-colorInherit",
                  "colorPrimary": "MuiIconButton-colorPrimary",
                  "colorSecondary": "MuiIconButton-colorSecondary",
                  "disabled": "Mui-disabled",
                  "edgeEnd": "MuiIconButton-edgeEnd",
                  "edgeStart": "MuiIconButton-edgeStart",
                  "label": "MuiIconButton-label",
                  "root": "MuiIconButton-root",
                  "sizeSmall": "MuiIconButton-sizeSmall",
                }
              }
              disableRipple={true}
            >
              <WithStyles(ForwardRef(ButtonBase))
                aria-label="search"
                centerRipple={true}
                className="MuiIconButton-root"
                disableRipple={true}
                disabled={false}
                focusRipple={true}
              >
                <ForwardRef(ButtonBase)
                  aria-label="search"
                  centerRipple={true}
                  className="MuiIconButton-root"
                  classes={
                    Object {
                      "disabled": "Mui-disabled",
                      "focusVisible": "Mui-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  disableRipple={true}
                  disabled={false}
                  focusRipple={true}
                >
                  <button
                    aria-label="search"
                    className="MuiButtonBase-root MuiIconButton-root"
                    disabled={false}
                    onBlur={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label"
                    >
                      <ForwardRef
                        color="inherit"
                      >
                        <WithStyles(ForwardRef(SvgIcon))
                          color="inherit"
                        >
                          <ForwardRef(SvgIcon)
                            classes={
                              Object {
                                "colorAction": "MuiSvgIcon-colorAction",
                                "colorDisabled": "MuiSvgIcon-colorDisabled",
                                "colorError": "MuiSvgIcon-colorError",
                                "colorPrimary": "MuiSvgIcon-colorPrimary",
                                "colorSecondary": "MuiSvgIcon-colorSecondary",
                                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                "root": "MuiSvgIcon-root",
                              }
                            }
                            color="inherit"
                          >
                            <svg
                              aria-hidden={true}
                              className="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                              />
                            </svg>
                          </ForwardRef(SvgIcon)>
                        </WithStyles(ForwardRef(SvgIcon))>
                      </ForwardRef>
                    </span>
                  </button>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(IconButton)>
          </WithStyles(ForwardRef(IconButton))>
          <Styled(WithStyles(ForwardRef(InputBase)))
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            inputProps={
              Object {
                "aria-autocomplete": "list",
                "aria-haspopup": "false",
                "aria-label": "Search…",
              }
            }
            placeholder="Search…"
            spellCheck="false"
            type="search"
          >
            <WithStyles(ForwardRef(InputBase))
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              className="hPyQul"
              inputProps={
                Object {
                  "aria-autocomplete": "list",
                  "aria-haspopup": "false",
                  "aria-label": "Search…",
                }
              }
              placeholder="Search…"
              spellCheck="false"
              type="search"
            >
              <ForwardRef(InputBase)
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className="hPyQul"
                classes={
                  Object {
                    "adornedEnd": "MuiInputBase-adornedEnd",
                    "adornedStart": "MuiInputBase-adornedStart",
                    "colorSecondary": "MuiInputBase-colorSecondary",
                    "disabled": "Mui-disabled",
                    "error": "Mui-error",
                    "focused": "Mui-focused",
                    "formControl": "MuiInputBase-formControl",
                    "fullWidth": "MuiInputBase-fullWidth",
                    "input": "MuiInputBase-input",
                    "inputAdornedEnd": "MuiInputBase-inputAdornedEnd",
                    "inputAdornedStart": "MuiInputBase-inputAdornedStart",
                    "inputHiddenLabel": "MuiInputBase-inputHiddenLabel",
                    "inputMarginDense": "MuiInputBase-inputMarginDense",
                    "inputMultiline": "MuiInputBase-inputMultiline",
                    "inputTypeSearch": "MuiInputBase-inputTypeSearch",
                    "marginDense": "MuiInputBase-marginDense",
                    "multiline": "MuiInputBase-multiline",
                    "root": "MuiInputBase-root",
                  }
                }
                inputProps={
                  Object {
                    "aria-autocomplete": "list",
                    "aria-haspopup": "false",
                    "aria-label": "Search…",
                  }
                }
                placeholder="Search…"
                spellCheck="false"
                type="search"
              >
                <div
                  autoCapitalize="none"
                  autoCorrect="off"
                  className="MuiInputBase-root hPyQul"
                  onClick={[Function]}
                  spellCheck="false"
                >
                  <input
                    aria-autocomplete="list"
                    aria-haspopup="false"
                    aria-label="Search…"
                    autoComplete="off"
                    className="MuiInputBase-input MuiInputBase-inputTypeSearch"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Search…"
                    type="search"
                  />
                </div>
              </ForwardRef(InputBase)>
            </WithStyles(ForwardRef(InputBase))>
          </Styled(WithStyles(ForwardRef(InputBase)))>
        </div>
      </styled.div>
    </div>
  </styled.div>
</ForwardRef>
`;
